src/kaby_lake/plane.cpp

```diff
@@ -0,0 +1,143 @@
+#include <array>
+
+#include <lil/fourcc.h>
+#include <lil/imports.h>
+
+#include "src/helpers.h"
+#include "src/kaby_lake/plane.hpp"
+#include "src/kaby_lake/watermarks.hpp"
+#include "src/regs.h"
+
+namespace {
+
+void wait_for_vblank(LilGpu *gpu, LilPlane *plane) {
+	REG(IIR(plane->pipe_id)) |= 1;
+
+	if(!wait_for_bit_set(REG_PTR(IIR(plane->pipe_id)), 1, 500000, 1000))
+		lil_log(WARNING, "timeout on wait_for_vblank\n");
+}
+
+}
+
+namespace kbl::plane {
+
+bool update_primary_surface(struct LilGpu* gpu, struct LilPlane* plane, GpuAddr surface_address, GpuAddr line_stride) {
+	if(surface_address & 0xFFF)
+		return false;
+
+    REG(PRI_SURFACE(plane->pipe_id)) = surface_address;
+
+    return true;
+}
+
+void page_flip(LilGpu *gpu, LilCrtc *crtc) {
+	REG(PRI_SURFACE(crtc->pipe_id)) = REG(PRI_SURFACE(crtc->pipe_id));
+}
+
+namespace {
+
+uint32_t linetime_us(uint32_t pixel_clock, uint32_t htotal) {
+	return div_u32_ceil(htotal * 1000, pixel_clock);
+}
+
+}
+
+void enable(LilGpu *gpu, LilCrtc *crtc, bool vblank_wait) {
+	uint32_t htotal = crtc->current_mode.htotal;
+	uint32_t pixel_clock = crtc->current_mode.clock;
+
+	uint8_t wm[8];
+	wm[0] = (gpu->mem_latency_first_set >> 0) & 0xFF;
```

Comment:
The compiler should optimize these away, but the `& 0xFF` _should_ not be needed here, as far as i know, as the array that is written to is a `uint8_t`

Reply:
yes but it helps readability IMO

Reply:
understandable